nova/virt/vmwareapi/driver.py

```diff
@@ -454,60 +437,66 @@ def get_available_nodes(self, refresh=False):
+        # the resources of all the nodes via
+        #    get_inventory & get_available_resource
+        # We follow here the same pattern as in the ironic driver and use this
+        # function call as an indicator of a new polling cycle and refresh
+        # the host stats cached in _vc_state by calling...
+        self._vc_state.get_host_stats(refresh=True)
+        # In the following calls to get_inventory and get_available_resource
+        # for each node, we then return the cached data
         return [self._nodename]
 
     def get_inventory(self, nodename):
         """Return a dict, keyed by resource class, of inventory information for
         the supplied node.
         """
-        stats = vm_util.get_stats_from_cluster(self._session,
-                                               self._cluster_ref)
-        datastores = ds_util.get_available_datastores(self._session,
-                                                      self._cluster_ref,
-                                                      self._datastore_regex)
-        total_disk_capacity = sum([ds.capacity for ds in datastores])
-        max_free_space = max([ds.freespace for ds in datastores])
+        stats = self.get_available_resource(nodename)
         reserved_disk_gb = compute_utils.convert_mb_to_ceil_gb(
             CONF.reserved_host_disk_mb)
         result = {
             fields.ResourceClass.DISK_GB: {
-                'total': total_disk_capacity // units.Gi,
+                'total': stats["local_gb"],
                 'reserved': reserved_disk_gb,
                 'min_unit': 1,
-                'max_unit': max_free_space // units.Gi,
+                'max_unit': stats["local_gb_max_free"],
                 'step_size': 1,
             }
         }
-        if stats['cpu']['max_vcpus_per_host'] > 0:
-            reserved_vcpus = stats['cpu'].get('reserved_vcpus', 0)
-            reserved_vcpus += CONF.reserved_host_cpus
-            result.update({fields.ResourceClass.VCPU: {
-                'total': stats['cpu']['vcpus'],
-                'reserved': reserved_vcpus,
-                'min_unit': 1,
-                'max_unit': stats['cpu']['max_vcpus_per_host'],
-                'step_size': 1,
-            }})
-        if stats['mem']['max_mem_mb_per_host'] > 0:
-            reserved_memory_mb = stats['mem'].get('reserved_memory_mb', 0)
-            reserved_memory_mb += CONF.reserved_host_memory_mb
-            result.update({fields.ResourceClass.MEMORY_MB: {
-                'total': stats['mem']['total'],
-                'reserved': reserved_memory_mb,
+
+        reserved_vcpus = stats["vcpus_reserved"]
+        result.update({fields.ResourceClass.VCPU: {
+            'total': stats["vcpus"],
+            'reserved': reserved_vcpus,
+            'min_unit': 1,
+            'max_unit': stats.get("max_vcpus_per_host",
+                            stats['vcpus'] - stats["vcpus_reserved"]),
+            'step_size': 1,
```

Comment:
Thinking about the `if` we had in the code before ... can `max_unit` become negative or 0 in this scenario? Would that be valid? I would assume this would break when getting added to/updated in placement. Might be better to not report the class if there are no vcpus available currently, e.g. because all hosts are in maintenance during build-up. This would help, because we wouldn't get a constant stream of errors reported into sentry.

Reply:
Yes, there is a case when this could become negative.
`CONF.reserved_host_cpus` is added without checking the maximum of cpus (as it was in the code before). I've added a limit for that to the total number of cpus.
I've also re-added the check that both `total` and  `max_unit` are larger than 0, as zero is not a valid value for those attributes.

Finally, I changed the calculation of the `max_unit`. In case of the aggregated view of the whole cluster, we subtract the reserved value from the total and store the maximum value in `max_x_per_host`, which is reasonable in that case,
as that is the maximum we can freely give out.

But in the segregated view, we have no `per_host` attribute and I was defaulting to subtract the reserved from the total, assuming that if we reserve more of the resource, the maximum needs to be smaller.
However, looking at the code in ironic how they handle when a node has been claimed, they simply increase the reserved value, without changing the `max_unit`. So, I rather stay with that pattern as simplifies the question, "can it become negative?"

Reply:
Yes, there is a case when this could become negative.
`CONF.reserved_host_cpus` is added without checking the maximum of cpus (as it was in the code before). I've added a limit for that to the total number of cpus.
I've also re-added the check that both total and max_unit are larger than 0, as zero is not a valid value for those attributes.

Finally, I changed the calculation of the `max_unit`. In case of the aggregated view of the whole cluster, we subtract the reserved value from the total and store the maximum value in `max_x_per_host`, which is reasonable in that case,
as that is the maximum we can freely give out.

But in the segregated view, we have no `per_host` attribute and I was defaulting to subtract the reserved from the total, assuming that if we reserve more of the resource, the maximum needs to be smaller.
However, looking at the code in ironic how they handle when a node has been claimed, they simply increase the reserved value, without changing the max_unit. So, I rather stay with that pattern as simplifies the question, "can it become negative?"

Reply:
The `max_unit` is now inconsistently set depending on if the hosts are reported in cluster-view or outside of cluster-view. That doesn't make sense as we still can only provide the `vcpus - reserved` of a host, no matter if it's aggregated or not. It's also confusing, as we now can't see the actual number of supported vcpus anymore, but have to compute them by using `total - reserved` to see what biggest flavor could support.

The way it's currently written, you'd give out the whole number of vcpus and reserve nothing - does this add up? Isn't `max_unit` then higher than `total * overcommit - reserved` i.e. the "available vcpus" (with overcommit = 1 on a single host)? Does placement allow that?

How does libvirt do it? They are always reporting for a single host, which should be the same as our single-hypervisor.

I don't really understand what claiming i.e. usage has to do with reserving stuff. Claiming something should be an allocation entry and should not change reserved, max_unit or total.

Reply:
> The max_unit is now inconsistently set depending [...]

I modeled the non-aggregated case after [libvirt](https://github.com/sapcc/nova/blob/stable/rocky-m3/nova/virt/libvirt/driver.py#L6608-L6610). Placement is aware of not being able to give out more than (total - reservation) of a resource.
The `max_unit` has only been introduced for VMware, where we have an aggregated view and where it is possible that `max_unit < total`, and since we only give out an aggregated view, we have to subtract there the reservation per host.

We can either be consistent with how vmware was doing its special thing, or consistent with the libvirt (and "fake").

> The way it's currently written, you'd give out the whole number of vcpus and reserve nothing - does this add up? Isn't max_unit then higher than total * overcommit - reserved i.e. the "available vcpus" (with overcommit = 1 on a single host)? Does placement allow that?

I don't follow... the inventory only reports actual resources. It reports the "physical" vcpus per host, total and reserved, per host. Over-Commitment is broken out as an additional optional parameter [allocation_ration](https://docs.openstack.org/api-ref/placement/?expanded=update-resource-provider-detail,update-resource-provider-inventory-detail#update-resource-provider-inventories) ([libvirt](https://github.com/sapcc/nova/blob/stable/rocky-m3/nova/virt/libvirt/driver.py#L6612)).

> How does libvirt do it? They are always reporting for a single host, which should be the same as our single-hypervisor.

Well, that is partly impossible, as we do have a `max_unit  < total`. It will be possible if we expose it as per-esxi, which the code is a preparation for.

> I don't really understand what claiming i.e. usage has to do with reserving stuff. Claiming something should be an allocation entry and should not change reserved, max_unit or total.

I am not sure if I got your question here correctly. We indeed have to separate use-cases for querying the hosts. One is the usage for reporting metrics, and the other is reporting the inventory of the available hardware.
I have moved them together not because they affect one another, but because there is an overlap in the attributes to be queried (and aggregated), because both need the information about the "physical" available resources.
The usage ("RAM/vCPUs used") is not affecting the inventory.